IoReader enhancements

chrysn opened the pull request:
These patches are by-catch of my work on #79, but should can stand on their own; in summary:

* Add more test cases for indefinite length strings (to catch what I think would be the most likely to go wrong with the following changes)
* Simplify `.read_into()` by using a provided method of the Read trait that does mostly the same (what's left is EOF error wrapping)
* In `.read()`, use [`Read::take()`](https://doc.rust-lang.org/stable/std/io/trait.Read.html#method.take) (with [`Read::by_ref()`](https://doc.rust-lang.org/stable/std/io/trait.Read.html#method.by_ref) as in its example to keep the original reader in place) to directly put the read data into the buffer vector without zeroing out a fresh area before

I haven't run any benchmarks on them, but I expect the read_into change to result in the same machine under optimization, and the read change to negligibly remove a cheap memory clearing -- but the former reduces code complexity, and the latter uses the vec's length cursor (which is available anyway) to more clearly express which portion of the scratch buffer is usable.

pyfisch commented:
Looks good to me! But I found the code complicated to understand. Would mind to add some comments to the reader?

chrysn commented on src/read.rs:
```diff
@@ -110,49 +110,51 @@ where
-        mut scratch_offset: usize,
+        scratch_offset: usize,
     ) -> Result<Reference<'de>> {
+        assert!(scratch.len() == scratch_offset);
```
Yes, at least while the Deserializer uses it as it does now.

I haven't changed the internal API part yet because there are two ways of going forward here (drop scratch_offset, or drop the buf.clear() in the reader and make scratch_offset a boolean "start anew"), see https://github.com/pyfisch/cbor/pull/80/commits/d1903ffe2a0a30b66d693348fa4f4978f171d938 -- but I think those are better explored in a separate change.

sfackler replied:
Ok cool - it seems fine to follow up in a separate change.

chrysn commented:
I've pushed the requested additional comments along with a typo as fixups, which I'd squash once the PR is good to go otherwise (or earlier if you prefer that workflow).

There's an additional commit on top now that does away with the "to_read be only up to 16k" loop, because I've read up on the read_to_end function that's now being used, and that already does something similar internally. The requested size truncated to 16k is still requested upfront to utilize the knowledge we have of the expected size, but if it's actually more, read_to_end will do about as well as the loop had.

pyfisch commented:
Good work. I think this is good to go.

Do you have any more comments @sfackler?

sfackler commented on src/read.rs:
```diff
@@ -110,49 +110,46 @@ where
-            scratch_offset = scratch.len();
+        if let Some(ch) = self.ch.take() {
+            scratch.push(ch);
+            n -= 1;
```
Can we ever call this method with `n == 0`?

sfackler commented:
LGTM other than the possible need for a check for `n == 0`.

chrysn commented:
Squashed into sensible commits (ie. all fixups resolved).

The n==0 was a good point, it's now fixed, commented and tested for. (Though the peek-and-then-becomes-0 branch is something I wouldn't know how to trigger w/o testing deep into the internals.)

chrysn commented:
Rebasing the other branch, I noted that the promised fix for n == 0 got lost somehow during the squashing or merging; I've picked it from the reflogs and have it as the first commit in the next series upcoming for #81 (where I'm ironing out a few readability issues).
